pkg/controller/stacks/stack/stack.go

```diff
@@ -685,30 +685,53 @@ func (h *stackHandler) prepareHostAwareJob(j *batch.Job, tokenSecret string) err
 	return nil
 }
 
-func (h *stackHandler) processDeployment(ctx context.Context) error {
+func (h *stackHandler) prepareDeployment() *apps.Deployment {
 	controllerDeployment := h.ext.Spec.Controller.Deployment
 	if controllerDeployment == nil {
 		return nil
 	}
 
-	// ensure the deployment is set to use this stack's service account that we created
+	// force the deployment to use stack opinionated names and service account
+	name := h.ext.Name + "-controller"
+	matchLabels := map[string]string{"app": name}
 	deploymentSpec := *controllerDeployment.Spec.DeepCopy()
 	deploymentSpec.Template.Spec.ServiceAccountName = h.ext.Name
+	deploymentSpec.Template.SetName(name)
+	dtLabels := deploymentSpec.Template.GetLabels()
+	if dtLabels == nil {
+		dtLabels = map[string]string{}
+	}
+	for k, v := range matchLabels {
+		dtLabels[k] = v
+	}
+	deploymentSpec.Template.SetLabels(dtLabels)
+	deploymentSpec.Selector = &metav1.LabelSelector{MatchLabels: matchLabels}
 
 	labels := stacks.ParentLabels(h.ext)
-	gvk := schema.GroupVersionKind{
-		Group:   apps.GroupName,
-		Kind:    "Deployment",
-		Version: apps.SchemeGroupVersion.Version,
-	}
-	d := &apps.Deployment{
+
+	return &apps.Deployment{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      controllerDeployment.Name,
+			Name:      name,
 			Namespace: h.ext.Namespace,
 			Labels:    labels,
 		},
 		Spec: deploymentSpec,
 	}
+}
+
+func (h *stackHandler) processDeployment(ctx context.Context) error {
+	d := h.prepareDeployment()
```

Comment:
It seems like it's related to [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) . . . in this case, `processDeployment` doesn't really care where the deployment comes from, so it probably shouldn't be calling into the method to generate the deployment. It would make it simpler in the future if we ever want to pass parameters into `prepareDeployment` from the code which is calling `processDeployment`

Reply:
I see how this could be modified to offer this.

`processDeployment` would send `prepareDeployment()` a bare `&appsv1.Deployment{}` which `prepareDeployment` would apply the `spec` using `DeepCopyInto`.

The `SetName` and `SetNamespace` would be used and the original labels would be merged with the parent labels instead of clobbered.

Reply:
As far as making `processDeployment` take parameters, that would warrant an out of scope refactor of the Stack controller's `create` function: https://github.com/crossplaneio/crossplane/blob/4534223ff95e7fb1f3ff559a792d65f49d0b855e/pkg/controller/stacks/stack/stack.go#L197-L229